datastore/datastore/storagebackend/postgresql/getobservations.go

```diff
@@ -219,100 +331,199 @@ func getGeoFilter(inside *datastore.Polygon, phVals *[]interface{}) (string, err
+func getObs(db *sql.DB, request *datastore.GetObsRequest, obs *[]*datastore.Metadata2) (retErr error) {
+
+	// get values needed for query
+	phVals := []interface{}{} // placeholder values
+	timeFilter, geoFilter, stringMdataFilter, err := createObsQueryVals(request, &phVals)
+	if err != nil {
+		return fmt.Errorf("createQueryVals() failed: %v", err)
+	}
+
+	// define and execute query
 	query := fmt.Sprintf(`
-		SELECT ts_id, observation.id, geo_point_id, pubtime, data_id, history, metadata_id,
-			obstime_instant, processing_level, value, point
+		SELECT
+		    ts_id,
+			obstime_instant,
+			pubtime,
+			value,
+			point,
+			%s
 		FROM observation
-		    JOIN geo_point gp ON observation.geo_point_id = gp.id
-			JOIN time_series ts on ts.id = observation.ts_id
-		WHERE %s AND %s AND %s AND %s
+		JOIN time_series on time_series.id = observation.ts_id
+		JOIN geo_point ON observation.geo_point_id = geo_point.id
+		WHERE %s AND %s AND %s
 		ORDER BY ts_id, obstime_instant
-	`, timeExpr, tsMdataExpr, obsMdataExpr, geoExpr)
+	`, strings.Join(obsStringMdataCols, ","), timeFilter, geoFilter, stringMdataFilter)
 
 	rows, err := db.Query(query, phVals...)
 	if err != nil {
 		return fmt.Errorf("db.Query() failed: %v", err)
 	}
 	defer rows.Close()
 
-	obsMap := make(map[int64][]*datastore.ObsMetadata)
+	obsMdatas := make(map[int64][]*datastore.ObsMetadata) // observations per time series ID
+
+	// scan rows
 	for rows.Next() {
-		var (
-			tsID            int64
-			id              string
-			gpID            int64
-			pubTime0        time.Time
-			dataID          string
-			history         string
-			metadataID      string
-			obsTimeInstant0 time.Time
-			processingLevel string
-			value           string
-			point           postgis.PointS
-		)
-		if err := rows.Scan(&tsID, &id, &gpID, &pubTime0, &dataID, &history, &metadataID,
-			&obsTimeInstant0, &processingLevel, &value, &point); err != nil {
-			return fmt.Errorf("rows.Scan() failed: %v", err)
+		obsMdata, tsID, err := scanObsRow(rows)
+		if err != nil {
+			return fmt.Errorf("scanObsRow() failed: %v", err)
 		}
 
-		obsMdata := &datastore.ObsMetadata{
-			Id: id,
-			Geometry: &datastore.ObsMetadata_GeoPoint{
-				GeoPoint: &datastore.Point{
-					Lon: point.X,
-					Lat: point.Y},
-			},
-			Pubtime:    timestamppb.New(pubTime0),
-			DataId:     dataID,
-			History:    history,
-			MetadataId: metadataID,
-			Obstime: &datastore.ObsMetadata_ObstimeInstant{
-				ObstimeInstant: timestamppb.New(obsTimeInstant0),
-			},
-			ProcessingLevel: processingLevel,
-			Value:           value,
-		}
-		obsMap[tsID] = append(obsMap[tsID], obsMdata)
+		obsMdatas[tsID] = append(obsMdatas[tsID], obsMdata)
 	}
 
 	// get time series
-	tsMdata := map[int64]*datastore.TSMetadata{}
+	tsMdatas := map[int64]*datastore.TSMetadata{}
 	tsIDs := []string{}
-	for id := range obsMap {
-		tsIDs = append(tsIDs, fmt.Sprintf("%d", id))
+	for tsID := range obsMdatas {
+		tsIDs = append(tsIDs, fmt.Sprintf("%d", tsID))
```

Comment:
I might be missing something... but if we have multiple observations from a time series, won't there by many duplicated in the `tdID`'s? This seem to make it all the way into the SQL query?

Probably easiest to make tsID's a "set"?

Reply:
I do think you missed something. A given tsID is processed only once since the loop iterates over the keys of obsMdatas which form a set by definition (just like the keys of any map).

Reply:
You are right of course. I don't know what my issue was. I'll blame it on it being 6pm.